skbio/diversity/alpha/_base.py

```diff
@@ -8,14 +8,118 @@
 
 from __future__ import absolute_import, division, print_function
 
+from functools import partial
+
 import numpy as np
+import pandas as pd
 from scipy.special import gammaln
 from scipy.optimize import fmin_powell, minimize_scalar
 
+import skbio
 from skbio.stats import subsample_counts
 from skbio.util._decorator import experimental
 from skbio.diversity._base import (_validate_counts_vector,
-                                   _validate_otu_ids_and_tree)
+                                   _validate_otu_ids_and_tree,
+                                   _vectorize_counts_and_tree)
+
+
+# need to figure out how to better factor this. as it stands, it's a big
+# duplication of .__init__.__all__
+def _get_alpha_diversity_metrics():
+    return {'ace': skbio.diversity.alpha.ace,
+            'chao1': skbio.diversity.alpha.chao1,
+            'chao1_ci': skbio.diversity.alpha.chao1_ci,
+            'berger_parker_d': berger_parker_d, 'brillouin_d': brillouin_d,
+            'dominance': dominance, 'doubles': doubles, 'enspie': enspie,
+            'esty_ci': esty_ci, 'faith_pd': faith_pd,
+            'fisher_alpha': fisher_alpha, 'goods_coverage': goods_coverage,
+            'heip_e': heip_e, 'kempton_taylor_q': kempton_taylor_q,
+            'margalef': margalef, 'mcintosh_d': mcintosh_d,
+            'mcintosh_e': mcintosh_e, 'menhinick': menhinick,
+            'michaelis_menten_fit': michaelis_menten_fit,
+            'observed_otus': observed_otus, 'osd': osd, 'pielou_e': pielou_e,
+            'robbins': robbins, 'shannon': shannon, 'simpson': simpson,
+            'simpson_e': simpson_e, 'singles': singles, 'strong': strong,
+            'gini_index': skbio.diversity.alpha.gini_index,
+            'lladser_pe': skbio.diversity.alpha.lladser_pe,
+            'lladser_ci': skbio.diversity.alpha.lladser_ci
+            }
+
+
+@experimental(as_of="0.4.0")
+def alpha_diversity(metric, counts, ids=None, **kwargs):
+    """ Compute alpha diversity for one or more count vectors
+
+    Parameters
+    ----------
+    metric: str, callable
+        The alpha diversity metric to apply to the count vector(s).
+        Passing metric as a string is preferable as this often results in an
+        optimized version of the metric being used.
+    counts : 1D or 2D array_like of ints or floats
+        Vector or matrix containing count/abundance data. If a matrix, each row
+        should contain counts of observations in a given sample.
+    ids : iterable of strs, optional
+        Identifiers for each sample in ``counts``.
+
+    Returns
+    -------
+    pd.Series
+        Values of metric for all vectors provided in ``counts``. The index
+        will be ``ids``, if provided.
+
+    Raises
+    ------
+    ValueError
+        If ``len(ids) != len(counts)``, or if ``otu_ids`` and ``tree`` are not
```

Comment:
List errors separately. 

*or if ``otu_ids``* to *if required metric-specific parameters are not provided*.

Add TypeError if invalid method-specific parameters are provided (and add a test of this).